### PR TITLE
Refs [TASK][36313] :Bind data : Get data pour into Recycler view coin

### PR DESCRIPTION
**Pull request by hideoncalm**
## Related Tickets
- [#Task 36313: Bind data : Get data pour into Recycler view coin](https://edu-redmine.sun-asterisk.vn/issues/36313)
- [#Task 36313: Bind data : Get data pour into Recycler view exchange](https://edu-redmine.sun-asterisk.vn/issues/36314)
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://github.com/framgia/coding-standards/tree/master/eng/android |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/coinqapp/base/BaseAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package com.sunasterisk.coinqapp.base
+
+import androidx.recyclerview.widget.RecyclerView
+
+abstract class BaseAdapter<T, V : BaseViewHolder<T>> : RecyclerView.Adapter<V>() {
+
+    private val items = mutableListOf<T>()
+
+    override fun onBindViewHolder(holder: V, position: Int) {
+        getItem(position)?.let { holder.onBindData(it) }
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/base/BaseAdapter.kt]: sao u ko lấy trực tiếp luôn u mà phải thông qua 1 hàm khác

File: app/src/main/java/com/sunasterisk/coinqapp/base/BaseViewHolder.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+abstract class BaseViewHolder<T>(binding: ViewBinding, onItemClick: (T) -> Unit) :
+    RecyclerView.ViewHolder(binding.root) {
+
+    private var itemData: T? = null
+
+    init {
+        itemView.setOnClickListener {
+            itemData?.let(onItemClick)
+            Toast.makeText(binding.root.context,"$itemData", Toast.LENGTH_LONG).show()
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/base/BaseViewHolder.kt]: bỏ check code đi nhé

File: app/src/main/java/com/sunasterisk/coinqapp/ui/home/HomeFragment.kt [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 import com.sunasterisk.coinqapp.base.BaseFragment
 import com.sunasterisk.coinqapp.databinding.FragmentHomeBinding
+import com.sunasterisk.coinqapp.ui.home.adapter.HomePagerAdapter
 
 class HomeFragment : BaseFragment<FragmentHomeBinding>() {
 
     override val binding by lazy { FragmentHomeBinding.inflate(layoutInflater) }
 
     override fun initViews() {
+        val adapter = HomePagerAdapter(childFragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT)
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/home/HomeFragment.kt]: có cần tách flag riêng ko vậy?

File: app/src/main/java/com/sunasterisk/coinqapp/ui/home/HomeFragment.kt [hunk 4]
```diff
@@ -1,13 +1,18 @@
...
+        binding.viewPager.adapter = adapter
+        binding.tabLayout.setupWithViewPager(binding.viewPager)
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/home/HomeFragment.kt]: dùng scope funtion nhé

File: app/src/main/java/com/sunasterisk/coinqapp/ui/home/adapter/HomePagerAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import androidx.fragment.app.FragmentPagerAdapter
+import com.sunasterisk.coinqapp.ui.listcoin.ListCoinFragment
+import com.sunasterisk.coinqapp.ui.listexchange.ListExchangeFragment
+
+class HomePagerAdapter(fm: FragmentManager, behavior: Int) : FragmentPagerAdapter(fm, behavior) {
+
+    override fun getCount() = 2
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/home/adapter/HomePagerAdapter.kt]: hard code

File: app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/ListCoinFragment.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    override fun hideLoading() {
+    }
+
+    private fun itemCoinClick(coin: Coin) {
+    }
+}
+
+
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/ListCoinFragment.kt]: thừa dòng

File: app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/adapter/ListCoinAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            }
+        }
+        companion object{
+            const val PATTEN_TEX_DAY = "#0.00%"
+            const val PATTEN_TEXT_PRICE = "$###,###.####"
+            const val PATTEN_TEXT_MARKET = "$###,###,##0"
+        }
+    }
+}
+
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/adapter/ListCoinAdapter.kt]: format code nhé, khoảng trắng

File: app/src/main/java/com/sunasterisk/coinqapp/ui/listexchange/ListExchangeFragment.kt [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    override fun showLoading() {
+    }
+
+    override fun hideLoading() {
+    }
+
+    private fun onItemClick(exchange: Exchange) {
+        context?.showMessage(exchange.id)
+    }
+
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/listexchange/ListExchangeFragment.kt:47]: .

File: app/src/main/java/com/sunasterisk/coinqapp/ui/listexchange/adapter/ListExchangeAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import java.text.DecimalFormat
+
+class ListExchangeAdapter(private val onItemClick: (Exchange) -> Unit) :
+    BaseAdapter<Exchange, ListExchangeAdapter.ListExchangeViewHolder>() {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListExchangeViewHolder {
+        val binding = ItemExchangeBinding.inflate(
+            LayoutInflater.from(parent.context),
+            parent, false
+        )
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/listexchange/adapter/ListExchangeAdapter.kt]:
```suggestion
        val binding = ItemExchangeBinding.inflate(
            LayoutInflater.from(parent.context),
            parent, 
            false
        )
```

File: app/src/main/java/com/sunasterisk/coinqapp/ui/home/HomeFragment.kt (hunk 3 above)
Comment (hideoncalm) [reply app/src/main/java/com/sunasterisk/coinqapp/ui/home/HomeFragment.kt]: nó là tham số truyền vào ý

File: app/src/main/java/com/sunasterisk/coinqapp/data/repository/ExchangeRepository.kt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    private val remote: ExchangeDataSource.Remote
+) : ExchangeDataSource.Remote {
+
+    override fun getExchanges(
+        perPage: Int,
+        page: Int,
+        callBack: OnLoadDataCallBack<List<Exchange>>
+    ) {
+        remote.getExchanges(perPage, page, callBack)
+    }
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/data/repository/ExchangeRepository.kt]:
```suggestion
    override fun getExchanges(
        perPage: Int,
        page: Int,
        callBack: OnLoadDataCallBack<List<Exchange>>
    ) = remote.getExchanges(perPage, page, callBack)
```

File: app/src/main/java/com/sunasterisk/coinqapp/ui/home/adapter/HomePagerAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import com.sunasterisk.coinqapp.ui.listexchange.ListExchangeFragment
+
+class HomePagerAdapter(fm: FragmentManager, behavior: Int) : FragmentPagerAdapter(fm, behavior) {
+
+    override fun getCount() = ITEM_COUNT
+
+    override fun getItem(position: Int): Fragment {
+        if (position == 0) return ListCoinFragment()
+        else return ListExchangeFragment()
+    }
```
Comment (trangnq258) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/home/adapter/HomePagerAdapter.kt]: nên tạo 1 adapter chứa listFragment và listTitle để tái sử dụng, tránh fix cứng fragment, title như này.

File: app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/ListCoinFragment.kt [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import com.sunasterisk.coinqapp.data.repository.CoinRepository
+import com.sunasterisk.coinqapp.data.source.remote.CoinRemoteDataSource
+import com.sunasterisk.coinqapp.databinding.FragmentListCoinBinding
+import com.sunasterisk.coinqapp.ui.listcoin.adapter.ListCoinAdapter
+import com.sunasterisk.coinqapp.utils.showMessage
+
+class ListCoinFragment : BaseFragment<FragmentListCoinBinding>(), ListCoinContract.View {
+
+    private var presenter: ListCoinContract.Presenter? = null
+    private val adapter = ListCoinAdapter(this::itemCoinClick)
```
Comment (trangnq258) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/ListCoinFragment.kt]:
```suggestion
    private val adapter = ListCoinAdapter(::itemCoinClick)
```

File: app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/adapter/ListCoinAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    BaseAdapter<Coin, ListCoinAdapter.ListCoinViewHolder>() {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListCoinViewHolder =
+        ListCoinViewHolder(
+            ItemCoinBinding.inflate(
+                LayoutInflater.from(parent.context),
+                parent,
+                false
+            ), onItemClick
+        )
```
Comment (trangnq258) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/adapter/ListCoinAdapter.kt]:
```suggestion
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        ListCoinViewHolder(
            ItemCoinBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            ), onItemClick
        )
```

File: app/src/main/java/com/sunasterisk/coinqapp/base/BaseAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+import androidx.recyclerview.widget.RecyclerView
+
+abstract class BaseAdapter<T, V : BaseViewHolder<T>> : RecyclerView.Adapter<V>() {
+
+    private val items = mutableListOf<T>()
+
+    override fun onBindViewHolder(holder: V, position: Int) {
+        items[position]?.let { holder.onBindData(it) }
+    }
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/base/BaseAdapter.kt]: ko cần check như vậy đâu truyền thẳng đc nhé nếu size =0 sẽ ko chạy vào đây đâu

File: app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/ListCoinFragment.kt [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+import com.sunasterisk.coinqapp.utils.showMessage
+
+class ListCoinFragment : BaseFragment<FragmentListCoinBinding>(), ListCoinContract.View {
+
+    private var presenter: ListCoinContract.Presenter? = null
+    private val adapter = ListCoinAdapter(::itemCoinClick)
+
+    override val binding by lazy {
+        FragmentListCoinBinding.inflate(layoutInflater)
+    }
```
Comment (phucsaker99) [inline app/src/main/java/com/sunasterisk/coinqapp/ui/listcoin/ListCoinFragment.kt]:
```suggestion
    override val binding by lazy { FragmentListCoinBinding.inflate(layoutInflater) }
```

**Pull request merged**